ion-schema/src/nfa.rs

```diff
@@ -4,135 +4,219 @@ use crate::types::TypeValidator;
-            let transition_to_states: &HashSet<Transition> =
-                self.nfa.transitions.get(transition_from_state_id).unwrap();
-            self.to_states_transitions(
-                transition_from_state_id,
-                transition_to_states,
-                can_exit,
-                visits,
-                event.to_owned(),
-                type_store,
-                &mut new_states,
-            );
+    pub fn validate_ordered_elements(&mut self, elements: Vec<Element>, type_store: &TypeStore) {
+        let mut elements_iter = elements.iter().peekable();
+        // use nfa_evaluation for validation
+        while let Some(element) = elements_iter.next() {
+            let mut next_states = HashSet::new();
+            for nfa_run in self.visits.iter() {
+                let transitions: HashSet<Transition> = self.nfa.get_transitions(nfa_run.state_id);
+                self.evaluate_transitions(
+                    nfa_run,
+                    transitions,
+                    element,
+                    &mut elements_iter,
+                    type_store,
+                    &mut next_states,
+                );
+            }
+            self.visits = next_states;
         }
-        new_states
     }
 
-    fn to_states_transitions(
+    fn evaluate_transitions(
         &self,
-        transition_from_state_id: &usize,
-        transition_to_states: &HashSet<Transition>,
-        can_exit: bool,
-        visits: &usize,
-        event: Option<Event>,
+        nfa_run: &NfaRun,
+        transitions: HashSet<Transition>,
+        element: &Element,
+        elements: &mut Peekable<Iter<Element>>,
         type_store: &TypeStore,
-        new_states: &mut HashMap<usize, usize>,
+        next_states: &mut HashSet<NfaRun>,
     ) {
-        let mut transition_to_states_for_optional_state: HashSet<Transition> = HashSet::new();
-        for transition_to_state in transition_to_states {
-            let transition_to_state_id = transition_to_state.destination;
-
-            if transition_to_state_id == *transition_from_state_id {
-                if self.nfa.states[transition_to_state_id].can_reenter(&(visits + 1)) {
-                    let can_enter = self.nfa.states[transition_to_state_id].can_enter(
-                        event.to_owned(),
-                        transition_to_state.type_id,
-                        type_store,
-                    );
-                    if can_enter {
-                        new_states.insert(transition_to_state_id, visits + 1);
+        let mut element = element;
+        let mut destination_states_for_optional_state: HashSet<Transition> = HashSet::new();
+
+        let source_state_id = nfa_run.state_id;
+        let visits = nfa_run.state_visits;
+
+        // traverse all transitions for source state and verify if we can take the given transition
+        for transition in transitions {
```

Comment:
This method is relatively complex (which is fine, this is the heart of the NFA logic!), but remember that future readers may not know what an NFA is or does. I think it'd be good to break this method into some smaller helper methods so it's easier to see the three main cases it tries to address:
1. Transition loops back to the same state
2. Transition moves to the next state
3. Transition moves to an optional state